lib/THC/THCGeneral.c

```diff
@@ -23,6 +23,22 @@ void THCudaInit(THCState* state)
   THCudaCheck(cudaGetDevice(&device));
 
   state->rngState = (THCRNGState*)malloc(sizeof(THCRNGState));
+
+#ifdef USE_CNMEM
+  cnmemDevice_t devices[count];
+  double multiplier = (getenv("CUTORCH_CNMEM")) ? (atof(getenv("CUTORCH_CNMEM"))) : (0.95);
+  for (int i = 0; i < count; ++i) {
+    devices[i].device = i;
+    size_t free_mem, used_mem;
+    cudaMemGetInfo(&free_mem, &used_mem);
+
+    devices[i].size = free_mem * multiplier;
+    devices[i].numStreams = 0;
+    devices[i].streams = NULL;
+  }
+  THCnmemCheck(cnmemInit(count, devices, CNMEM_FLAGS_DEFAULT));
```

Comment:
Yeah, it's set to 95% by default, but the user can configure it with an environment flag. This is the way it's done in Theano and was done in Caffe I believe (they now switched to another memory pool implementation). It can be problematic when people share GPUs, yes.

I haven't looked at it in detail, but I think CNMeM has support for growing built in. From the docstring of `cnmemMalloc`: 

> If there's a buffer of size larger or equal to the requested size in the list of free blocks, it is returned. If there's no such buffer but the manager is allowed to grow its memory usage (the CNMEM_FLAGS_CANNOT_GROW flag is not set), the memory manager calls cudaMalloc.

I have no idea if and how much of a performance penalty you'd incur by growing memory instead of pre-allocating.